python/ray/dag/compiled_dag_node.py

```
@@ -1312,65 +1314,32 @@ def _get_or_compile(
-                        # the InputNode.
-
-                        # TODO(jhumphri): Handle case where there is an actor,
-                        # other than just the driver actor, also reading the
-                        # output from the `task` node. For example, the following
-                        # currently does not work:
-                        # def test_blah(ray_start_regular):
-                        #     a = Actor.remote(0)
-                        #     b = Actor.remote(10)
-                        #     with InputNode() as inp:
-                        #         x = a.inc.bind(inp)
-                        #         y = b.inc.bind(x)
-                        #         dag = MultiOutputNode([x, y])
-
-                        #     compiled_dag = dag.experimental_compile()
-                        #     output_channel = compiled_dag.execute(1)
-                        #     result = output_channel.read()
-                        #     print(result)
-
-                        #     compiled_dag.teardown()
-
-                        assert self._proxy_actor is not None
-                        for reader in readers:
-                            reader_and_node_list.append(
+                    # Use reader_handles_set to deduplicate readers on the
+                    # same actor, because with CachedChannel each actor will
+                    # only read from the upstream channel once.
+                    reader_handles_set = set()
+                    read_by_multi_output_node = False
+                    for reader in readers:
+                        if isinstance(reader.dag_node, MultiOutputNode):
+                            assert self._proxy_actor is not None
+                            # inserting at 0 because CompositeChannel
+                            # _get_self_actor_id() reads first element for proxy actor
+                            reader_and_node_list.insert(
```

We also need `reader_handles_set` to dedup in this case? Can you add a unit test that has `dag = MultiOutputNode(x, x)`, i.e., driver reading the same channel multiple times?

---

I unified the dedup logic, and tried the case, but it fails because channel times out waiting. This is the same thing that happens on master right now. I think it's a separate problem, will take an initial look at it, but I think it might be a something else.

Should we raise an error right now when compiling the graph that says that the user can't read the same thing twice from the driver?